package/collection2/collection2.js

```diff
@@ -46,66 +46,66 @@ Mongo.Collection.prototype.attachSchema = function c2AttachSchema(ss, options) {
     ss = new SimpleSchema(ss);
   }
 
-  this._c2 = this._c2 || {};
-
-  // If we've already attached one schema, we combine both into a new schema unless options.replace is `true`
-  if (this._c2._simpleSchema && options.replace !== true) {
-    if (ss.version >= 2) {
-      var newSS = new SimpleSchema(this._c2._simpleSchema);
-      newSS.extend(ss);
-      ss = newSS;
-    } else {
-      ss = new SimpleSchema([this._c2._simpleSchema, ss]);
-    }
-  }
-
-  var selector = options.selector;
-
   function attachTo(obj) {
-    if (typeof selector === "object") {
+    // we need an array to hold multiple schemas
+    // position 0 is reserved for the "base" schema
+    obj._c2 = obj._c2 || {};
+    obj._c2._simpleSchemas = obj._c2._simpleSchemas || [ null ];
+
+    if (typeof options.selector === "object") {
+      // Selector Schemas
+
+      // Extend selector schema with base schema
+      var baseSchema = obj._c2._simpleSchemas[0];
+      if (baseSchema) {
+        ss = extendSchema(baseSchema.schema, ss);
+      }
+
       // Index of existing schema with identical selector
-      var schemaIndex = -1;
+      var schemaIndex;
 
-      // we need an array to hold multiple schemas
-      obj._c2._simpleSchemas = obj._c2._simpleSchemas || [];
+      // Loop through existing schemas with selectors,
+      for (schemaIndex = obj._c2._simpleSchemas.length - 1; 0 < schemaIndex; schemaIndex--) {
+        var schema = obj._c2._simpleSchemas[schemaIndex];
+        if (schema && isEqual(schema.selector, options.selector)) break;
+      }
 
-      // Loop through existing schemas with selectors
-      obj._c2._simpleSchemas.forEach((schema, index) => {
-        // if we find a schema with an identical selector, save it's index
-        if(isEqual(schema.selector, selector)) {
-          schemaIndex = index;
-        }
-      });
-      if (schemaIndex === -1) {
+      if (schemaIndex <= 0) {
```

Comment:
The less or equal to is really just a programming courtesy. In reality it could equally read `schemaIndex == 0`.

Because at line 68 we search backwards through through the existing selector schemas, and stop when `0 < schemaIndex`, once schemaIndex has reached zero from array end, without matching any previous selectors, we can be confident we can push the new selector onto the array.

The first point "When a schema with no selector is attached, either set, merge, or replace defaultSchema." is handled completely separately. Please see line 55 `if (typeof options.selector === "object") {` where the code referenced only handles attaching selector schemas.

Lines 90 through 108 handle point two.

https://github.com/rcsandell/meteor-collection2/blob/e463f51451394319bf098bd1062bce04b85f48a3/package/collection2/collection2.js#L90-L108

Reply:
The reason I went with storing everything in one array is that it makes it super easy to just loop through and check all schemas in the array, no special cases.